### PR TITLE
express-notes-api | Lucca Rizzi

### PR DESCRIPTION
**Pull request by luccarizzi**
new1

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+})
+
+app.get('/api/notes/:id', checkIntSign, (req, res) => {
+  const id = parseInt(req.params.id);
+  for (let note in notes) {
+    if (parseInt(note) === id) {
+      res.status(200).json(notes[id]);
+      return;
+    }
+  }
```
Comment (scbowler) [inline express-notes-api/index.js:26]: You do not need a loop here, you can directly check if it exists or not. For example:

```JAVASCRIPT
if(notes[id]) {
  // Note exists, send it to the client
} else {
  // Note does not exist, send a 404 error to the client
}
```

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+function checkIntSign(req, res, next) {
+  if (Math.sign(req.params.id) === -1) {
+    res.status(400).json({ 'error': 'id must be a positive integer' });
+    return;
+  } else if (isNaN(req.params.id)) {
+    res.status(400).json({ 'error': 'id must be a positive integer' });
+  } else {
+    next()
+  }
+}
```
Comment (scbowler) [inline express-notes-api/index.js:39]: This is great! Nice use of a middleware!!!!! Only thing is I would move this up towards the top of the file (above all endpoints). While it doesn't really matter it is just out of place here.

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      res.status(500).json({ "error": "An unexpected error occurred." });
+    } else {
+      res.status(201).json(req.body);
+    }
+  })
+  dataJson.nextId++;
+})
+
+function checkContent(req, res, next) {
+  if (Object.keys(req.body).length === 0) {
```
Comment (scbowler) [inline express-notes-api/index.js:57]: You don't need to check if the body has any properties, you just need to check if it has a `content` property

```JAVASCRIPT
if(!req.body.content) {
  // Missing content, send client an error
```

File: express-notes-api/index.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    res.status(400).json({ 'error': 'content is a required field' });
+    return;
+  } else {
+    next();
+  }
+}
```
Comment (scbowler) [inline express-notes-api/index.js:63]: Again, nice use of middleware, but you should move this up at the top.

File: express-notes-api/index.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+app.delete('/api/notes/:id', checkIntSign, checkId, (req, res) => {
+  delete notes[req.params.id];
```
Comment (scbowler) [inline express-notes-api/index.js:66]: You should verify if the note exists before deleting it, so if it doesn't exist you can send the client a 404 error

File: express-notes-api/index.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  })
+})
+
+function checkId (req, res, next) {
+  const id = parseInt(req.params.id);
+  for (let note in notes) {
+    if (parseInt(note) === id) {
+      return next();
+    }
+  }
```
Comment (scbowler) [inline express-notes-api/index.js:83]: You don't need a loop here just directly check if the property exists.

```JAVASCRIPT
if(notes[id]) {
  return next();
}

// send error
```

File: express-notes-api/index.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  return res.status(404).json({'error': `cannot find note with id ${id}`});
+}
```
Comment (scbowler) [inline express-notes-api/index.js:85]: I would group all of your middleware together at the top of the file, above your endpoints

**Review by scbowler — CHANGES_REQUESTED** [review]
Looking real good, and I LOVE the use of middleware! However there are a few things that need cleaned up.

File: express-notes-api/index.js (hunk 2 above)
Comment (luccarizzi) [reply express-notes-api/index.js]: Done.

File: express-notes-api/index.js (hunk 4 above)
Comment (luccarizzi) [reply express-notes-api/index.js]: Done.

File: express-notes-api/index.js (hunk 7 above)
Comment (luccarizzi) [reply express-notes-api/index.js]: Done. I also created a object to hold all middleware as methods and destructured the object.

File: express-notes-api/index.js (hunk 1 above)
Comment (luccarizzi) [reply express-notes-api/index.js]: Done.

File: express-notes-api/index.js (hunk 3 above)
Comment (luccarizzi) [reply express-notes-api/index.js]: Done.

File: express-notes-api/index.js (hunk 6 above)
Comment (luccarizzi) [reply express-notes-api/index.js]: Done.

File: express-notes-api/index.js (hunk 5 above)
Comment (luccarizzi) [reply express-notes-api/index.js:76]: There is `checkId` to check if the note exists. If it doesn't, checkId will send the client an error.

**Review by Zhiwutian — APPROVED** [review]
👍

**Pull request merged**